onnxruntime/test/python/transformers/test_skip_layer_norm_fusion.py

```diff
@@ -0,0 +1,276 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation.  All rights reserved.
+# Licensed under the MIT License.  See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import os
+import unittest
+from typing import Dict, List
+
+import numpy as np
+import onnx
+from onnx import TensorProto, helper, numpy_helper
+from parity_utilities import find_transformers_source
+
+if find_transformers_source():
+    from fusion_options import FusionOptions
+    from optimizer import optimize_model
+else:
+    from onnxruntime.transformers.fusion_options import FusionOptions
+    from onnxruntime.transformers.optimizer import optimize_model
+
+
+def float_tensor(name: str, shape: List[int], random=False):
+    low = 0.0
+    high = 1.0
+    total_elements = 1
+    for x in shape:
+        total_elements *= x
+    weights = [np.random.uniform(low, high) for _ in range(total_elements)] if random else [1.0] * total_elements
+    return helper.make_tensor(name, TensorProto.FLOAT, shape, weights)
+
+
+class TestFusion(unittest.TestCase):
+    def verify_skip_layer_norm_fusion(
+        self,
+        model_path: str,
+        expected_counter: Dict[str, int],
+        expected_inputs: List[str],
+        expected_outputs: List[str],
+    ):
+        options = FusionOptions("bert")
+        optimized_model = optimize_model(model_path, optimization_options=options, opt_level=0)
+
+        ops = ["Add", "LayerNormalization", "SkipLayerNormalization", "Cast"]
+        for op in ops:
+            nodes = optimized_model.get_nodes_by_op_type(op)
+            print(op, len(nodes), expected_counter[op])
+            self.assertEqual(len(nodes), expected_counter[op])
+
+            if op == "SkipLayerNormalization" and expected_counter[op] == 1:
+                print(nodes[0].input)
+                print(nodes[0].output)
+                self.assertEqual(nodes[0].input, expected_inputs)
+                self.assertEqual(nodes[0].output, expected_outputs)
+
+    def create_test_model(
+        self,
+        batch_size: int = 1,
```

Comment:
Can you add a test case with `batch_size > 1`? It will help test different cases for the shape of the `skip` input.

https://github.com/microsoft/onnxruntime/blob/47fe7fe90091e66fce7c6ef46d298ce610de359a/onnxruntime/core/graph/contrib_ops/bert_defs.cc#L1100

Reply:
Current fusion does not support broadcasting yet so only the first case can be fused. See https://github.com/microsoft/onnxruntime/blob/1aa62a5c7c0590d76b354eaa5c08e9d5e6068de8/onnxruntime/python/tools/transformers/fusion_skiplayernorm.py#L60-L67

We can add such a case when we add the support of broadcasting in fusion.